lte/gateway/c/core/oai/test/sgw_s8_task/sgw_csr.cpp

```diff
@@ -0,0 +1,354 @@
+  uint64_t imsi64 = 1010000000001;
+  virtual void SetUp() {
+    config->itti_config.queue_size     = 0;
+    std::string file_string            = "/var/opt/magma/tmp/spgw.conf";
+    config->itti_config.log_file       = bfromcstr(file_string.c_str());
+    std::string s1u_if_name            = "eth1";
+    config->ipv4.if_name_S1u_S12_S4_up = bfromcstr(s1u_if_name.c_str());
+    config->ipv4.S1u_S12_S4_up.s_addr  = 0x8e3ca8c0;
+    config->ipv4.netmask_S1u_S12_S4_up = 24;
+    std::string s5s8u_if_name          = "eth0";
+    config->ipv4.if_name_S5_S8_up      = bfromcstr(s5s8u_if_name.c_str());
+    config->ipv4.S5_S8_up.s_addr       = 0xf02000a;
+    config->ipv4.netmask_S5_S8_up      = 24;
+    std::string s11                    = "lo";
+    config->ipv4.if_name_S11           = bfromcstr(s11.c_str());
+    config->ipv4.S11.s_addr            = 0x100007f;
+    config->ipv4.netmask_S11           = 8;
+    config->udp_port_S1u_S12_S4_up     = 2152;
+    config->config_file                = bfromcstr(file_string.c_str());
+  }
+  virtual void TearDown() {
+    bdestroy_wrapper(&config->itti_config.log_file);
+    bdestroy_wrapper(&config->ipv4.if_name_S1u_S12_S4_up);
+    bdestroy_wrapper(&config->ipv4.if_name_S5_S8_up);
+    bdestroy_wrapper(&config->ipv4.if_name_S11);
+    bdestroy_wrapper(&config->config_file);
+    free(config);
+  }
+};
+
+// TC validates creation of UE context on reception of Create Session Req
+TEST_F(SgwS8Config, create_ue_context_cs_req) {
+  sgw_state_init(false, config);
+  sgw_state_t* sgw_state = get_sgw_state(false);
+
+  spgw_ue_context_t* ue_context_p = NULL;
+  mme_sgw_tunnel_t sgw_s11_tunnel = {0};
+  sgw_s11_tunnel.local_teid       = sgw_s8_generate_new_cp_teid();
+  EXPECT_EQ(
+      sgw_update_teid_in_ue_context(
+          sgw_state, imsi64, sgw_s11_tunnel.local_teid),
+      RETURNok);
+  EXPECT_EQ(
+      hashtable_ts_get(
+          sgw_state->imsi_ue_context_htbl, (const hash_key_t) imsi64,
+          reinterpret_cast<void**>(&ue_context_p)),
+      HASH_TABLE_OK);
+
+  sgw_state_exit();
+}
+
+// TC validates creation of pdn session on reception of Create Session Req
+TEST_F(SgwS8Config, create_pdn_session_cs_req) {
+  sgw_state_init(false, config);
+  sgw_state_t* sgw_state = get_sgw_state(false);
+
+  spgw_ue_context_t* ue_context_p = NULL;
+  mme_sgw_tunnel_t sgw_s11_tunnel = {0};
+  sgw_s11_tunnel.local_teid       = sgw_s8_generate_new_cp_teid();
+
+  EXPECT_EQ(
+      sgw_update_teid_in_ue_context(
+          sgw_state, imsi64, sgw_s11_tunnel.local_teid),
+      RETURNok);
+  EXPECT_EQ(
+      hashtable_ts_get(
+          sgw_state->imsi_ue_context_htbl, (const hash_key_t) imsi64,
+          reinterpret_cast<void**>(&ue_context_p)),
+      HASH_TABLE_OK);
+
+  sgw_eps_bearer_context_information_t* sgw_pdn_session = NULL;
+  sgw_pdn_session = sgw_create_bearer_context_information_in_collection(
+      sgw_s11_tunnel.local_teid);
+  EXPECT_TRUE(sgw_pdn_session != nullptr);
```

Comment:
since those test run only one single function, it may be good to check the values of the hashtables that modifies and the content of the `sgw_eps_bearer_context_information_t` list